### PR TITLE
[DO NOT MERGE, early full-scale unofficial changes validation only] Kimi K3: full tree with 5D parallelism multimodal, LoRA, MoonViT Dynamic CP&DEP, MXFP4/MXFP8 QAT, quantile balancing

### PR DESCRIPTION
**Pull request by QIU023**
Draft, and not for merge, providing for early timeline and accessable for full 2.8T param 5D smoke while waiting for PR-4025.

Would need maintainers thorough review on later major parallelize.py sharding.py declaration on major module parallelism support in titan.

Pending final evaluations and fixes before publishing: MXFP4 QAT and packed-MXFP4 import, quantile load balancing, MTP, LoRA, the decoupled vision encoder, a MoonEP token dispatcher, and the HF <-> DCP key map for the released checkpoint.

PR-4025 is a separate implementation of the same model and is further along on the model itself. When it lands, this rebases onto it and what remains is the parallelism and the post-training pieces. Filing now is disclosure, not a request to review 28k lines.

98 files, +28307 -28. The 28 deletions are the thing to notice -- this is almost entirely additive, and the five core files it does touch are listed in commit 14.

The fifteen commits are sliced by content rather than by the 409 commits of history that produced them:

| # | commit | main files |
| --- | --- | --- |
| 01 | the K3 model: KDA, MLA, latent MoE, MTP, and its config tree | `model.py` `model_configs.py` `config_registry.py` |
| 02 | Block Attention Residuals: the primitive and the layout tables | `attn_res.py` `attn_res_model.py` `layout.py` |
| 03 | SiTU-GLU routed experts and quantile load balancing | `moe.py` `quantile_balance.py` `common/moe.py` |
| 04 | MoonViT and the multimodal wrapper | `moonvit.py` `multimodal_model.py` `vision_preprocess.py` |
| 05 | the quantization scope -- what K3 quantizes, shared by QAT and QLoRA | `quant_scope.py` |
| 06 | context parallelism: KCP for KDA, Ulysses for MLA, dynamic CP for the tower | `kcp.py` `vit_cp_plan.py` |
| 07 | pipeline parallelism: the cross-stage AttnRes adapter | `pipeline_adapter.py` |
| 08 | the decoupled vision encoder, with bubble scheduling | `dep_bubble_*.py` `vit_prefetch.py` |
| 09 | a MoonEP token dispatcher against torchtitan's EP seam | `moon_ep_dispatcher.py` |
| 10 | MXFP4 QAT and packed-MXFP4 weight import | `mxfp4_qat.py` `packed_mxfp4.py` |
| 11 | LoRA, including the skip-edge gradients PP has to route | `lora.py` `muon.py` |
| 12 | HF <-> DCP conversion for the released key set | `hf_key_map.py` `state_dict_adapter.py` |
| 13 | the parallelize entry that applies all of the above, incl. the TP plan and the KDA DTensor shims | `parallelize.py` |
| 14 | the core changes the above needs | `distributed/{fsdp,utils}.py` `components/{lr_scheduler,optimizer}.py` |
| 15 | tests | `tests/` (56 files) |

Two upstream defaults have to be overridden to run it, both pinned in our gate rather than worked around in the model. `spmd_backend` needs `partial_dtensor`: under `spmd_types`, `fully_shard()` wants every parameter to already be a DTensor on the full SPMD mesh, and TP additionally trips `assert_type() does not support DTensor`, so supporting it is the declarative conversion itself rather than a patch on top -- that is the largest gap between this `parallelize.py` and an upstream model's, and it is what we are doing next. And CUDA graph capture needs `--training.disable-cuda-graphs`, because the vision path's patch count varies per batch and capture validation rejects it; that one is not specific to this model, so it may deserve its own issue.

Evidence: 58 of 58 gate cells pass on this tree, ten steps each, with those two pins. Per-cell loss and grad_norm for every step of all 58 are here, one line per cell:

https://github.com/QIU023/torchtitan_attention_residual/blob/f81a19319506ef1c3d3e2b8fc6eadbf7e6d99feb/phase13_k3like_48b_posttrain/gate_logs/gate_58_2026-08-21_final_percell.txt

The three model arms are text, multimodal, and multimodal+LoRA, 18 cells each, over dp / fsdp / tp / cp / pp / ep and their combinations up to `ep2_fsdp2_tp2_cp2`; the remaining four are the two decoupled-encoder cells at pp4 and pp8 and the two pp8xvp4 cells.

### CODE REVIEW

**Comment by elfiegg** [comment]
Here is a quick summary of progress from agents, please take a look at the functional issue section - larger scale validation is still undergoing.  

## GB200 Validated functionality

| Functionality | Registered configuration / artifact | Scale | Evidence | Status |
|---|---|---:|---|---|
| Release artifact integrity | Official Kimi-K3 checkpoint, 96 shards | CPU - File validation | All 497,220 index keys close; no unmapped or non-stacked runtime mapping mismatches | Passed |
| Basic text training | `kimi_k3_debugmodel_report_arch` | 1 GPU | Deterministic 10-step runs with identical TensorBoard scalar traces | Passed |
| KDA, MLA, routed MoE, AttnRes | `kimi_k3_debugmodel_report_arch` | 1 GPU | Finite loss and nonzero gradients in the principal KimiK3 components | Passed |
| Image + text training | `kimi_k3_debugmodel_report_arch` + MoonViT | 1 GPU | Image input changes logits; vision patch embedding receives nonzero gradient | Passed |
| LoRA fine-tuning | `kimi_k3_debugmodel_report_arch_lora` | 1 GPU | Base weights frozen, adapters update, merge error 0.0, HF export contains 604 tensors | Passed |
| QAT | `kimi_k3_debugmodel_report_arch_qat` | 1 GPU | Routed-expert MXFP4/MXFP8 fake quantization, finite STE gradients, 10-step training | Passed |
| Multi-token prediction | `kimi_k3_debugmodel_report_arch` + MTP loss | 1 GPU | Trainer-compatible `positions=` path and packed-document boundaries, 10 steps | Passed |
| Muon + Quantile Balancing | `kimi_k3_debugmodel_report_arch` | 1 GPU | Muon/AdamW resolution, QB hooks in 20 MoE layers, tag refresh, 10 recipe steps | Passed |
| Distributed parallelism | `kimi_k3_debugmodel_report_arch` | 4 GPUs / 1 node | TP2xCP2, FSDP2xTP2, and FSDP2xTP2xEP2 finite training runs | Passed |
| Direct Slurm multi-node preflight | `kimi_k3_debugmodel_report_arch` | 8 GPUs / 2 nodes | Direct rank setup, all-reduce/all-to-all, repeated FSDP2xTP2xCP2 10-step mini-run | Passed |
| PP4 text pipeline training | `kimi_k3_debugmodel_report_arch` | Up to 8 GPUs / 2 nodes | Ten pipeline training steps | Passed |
| PP4 multimodal pipeline training | `kimi_k3_debugmodel_report_arch` + MoonViT | Up to 8 GPUs / 2 nodes | Ten image+text pipeline training steps | Passed |
| Four-node initial pipeline training | `kimi_k3_debugmodel_report_arch` | 16 GPUs / 4 nodes | FSDP2xTP2xCP2xPP2 with EP2, two finite pipeline steps | Passed |

## Known functional issue
| Issue | Observed behavior | Resolution / next step | State |
|---|---|---|---|
| MTP trainer interface | `KimiMTPLoss` rejects trainer keyword `positions=` | Local draft fix accepts keyword-only `positions` and masks cross-document targets | Fixed locally; validated |
| Muon optimizer registration | Muon recipe fails before training because the active optimizer factory does not resolve Muon | Local draft fix registers/resolves Muon in the active factory | Fixed locally; validated |
| Muon tag lifetime | Parameter replacement during parallelization can discard per-head tags before optimizer grouping | Local draft fix refreshes tags immediately before groups are built | Fixed locally; validated |
| PP8 direct-Slurm startup | Text and multimodal PP8 reach all stages, then ranks 6/7 wait 300 s creating lazy NCCL P2P communicators for edges 5->6 and 6->7 | Run a minimal two-step PP8 NCCL diagnostic repro, then add an in-process eager PP-edge communicator warmup if confirmed | Open; launcher/runtime issue, not model numerics |
| AttnRes layer-0 gradient assertion | Layer 0 has no prior history, so its aggregation projection can have zero/absent gradient | Correct the test to assert a later layer rather than requiring every layer | Test-only correction |

## Remaining work (Pending GPU scheduling)

1. Run the 16-GPU quantization-aware training job and verify ten clean
   training steps, quantization activity, and complete per-rank metrics.
2. Run the 16-GPU image-and-text training job. Verify that the vision
   tower receives gradients, save a distributed checkpoint after step 5, reload
   it, and continue successfully through step 10. Then run the equivalent
   16-GPU LoRA job and verify frozen base weights and adapter updates.
3. Reproduce the eight-stage, two-node pipeline startup hang with one small
   direct-Slurm diagnostic. If it confirms the suspected NCCL P2P initialization
   ordering issue, add an eager pipeline-neighbor communicator warmup and rerun
   the affected text and image-and-text cases.
4. Complete the distributed-layout cases that did not run before the previous
   four-hour allocation expired. Submit only those missing cases in smaller,
   bounded allocations rather than repeating completed work.
5. Full 1.5T training - A complete production-scale plan
   for real checkpoint loading, memory budgeting, optimizer state, sharding,
   and a larger Slurm allocation. Deferred until initial performance optimizations are done.

## Current measured training performance on GB200
| Workload | Scale | Stable trainer-reported throughput | FLOPs | Peak memory per GPU | Loss over 10 steps |
|---|---:|---:|---:|---:|---:|
| Text training | 1 GB200 | ~2,200 tokens/s | ~0.88-0.90 TFLOP/s | 1.64 GiB | 7.72 -> 5.14 |
| Image + text training | 1 GB200 | ~1,330-1,440 tokens/s | ~0.54-0.59 TFLOP/s | 1.70 GiB | 12.05 -> 9.84 |
| Image + text LoRA | 1 GB200 | ~1,470-1,500 tokens/s | ~0.60-0.61 TFLOP/s | 1.02 GiB | 12.03 -> 11.90 |
| Image + text QAT | 1 GB200 | ~1,230-1,270 tokens/s | ~0.50-0.51 TFLOP/s | 1.73 GiB | 12.05 -> 9.86 |
| Image + text context-parallel test | 4 GB200s / 1 node | ~335-346 tokens/s | ~0.14 TFLOP/s | 1.51 GiB | 12.07 -> 9.89 |
| Image + text TP/PP/EP test | 8 GB200s / 2 nodes | ~104-112 tokens/s | ~0.04-0.05 TFLOP/s | 1.22-1.23 GiB | 12.07 -> 9.74 |

**Comment by QIU023** [comment]
Thank you so much Elfie! @elfiegg Please feel free to share your fork branch with some of the above fix if you have them locally to unblock your local runs, 

Specifically I am mostly concerning the "PP8 direct-Slurm startup" issue one because that is triggered in the critical multi-node PP setup and I do lack multi-node GPU development setup for now, which I definitely want to upgrade to. I have put a potential fix but found the fix might be located in `torch.distributed.pipelining` part due to lazy NCCL setup and the dramatic timing difference of the P2P NCCL stream launching in cross-node setup (your multi-node env) and intra-node setup (my running env), since the fix may be scoped in `pytorch` core distributed lib, not just torchtitan, we might want to discuss with @tianyu-l and @shuhuayu further on this.

I will share the potential fix commits since I already have them, but not sure if you are okay that I directly put it in this PR branch so you can pull down conveniently but please let me know if you found some breaking changes.

In the meantime, the post-rebase tree has migrated lots of changes and some part of them have became formal PRs: text LLM EP, CP, PP are already under review. Once we merged some of those, we can follow up to use the post-rebase tree OR the mainline merged PRs for K3 parallelism instead. This pre-rebase tree has most of the drafted features

**Comment by QIU023** [comment]
Hi Elfie @elfiegg thanks again, curious do you have any new evaluation outcomes ? (No rush/hard requirement at all, just follow up after a few weeks)

I had been busy on proceeding a post eager model rebased tree and iterating on formal PP, CP, EP (merged) and bunch of other features PRs raising and reviewing with titan maintainers, 

meanwhile I had https://github.com/pytorch/torchtitan/pull/4281/commits/d1ec535d1fdf6c8c9c931a341c205fb56069349c this commit pushed to potentially resolve the above mentioned PP NCCL hang issue, so latest branch HEAD of this PR can be retried on a multi-node PP setup.

Please keep me posted if you have any evaluation outcome or we can discuss with maintainers (cc @shuhuayu @tianyu-l ) to see if we want to run a full 2.8T run on a pretty landed titian tree soon or later (I have another post-rebase integration tree for K3, similar to this PR format), or LMK if you have any other ideas or upcoming plans, thanks!

For QAT MXFP4/FP8, I have seen lots of PRs from other contributers so feel free to check with those, I might not able to keep closest to QAT part given limited bandwidth.

**Comment by elfiegg** [comment]
Apologies for the delay! Perhaps we should set up a communication channel to share progress updates, blockers, and priorities.

With the updated branch, multi-node GB200 training still encounters the critical PP-related NCCL hang when using PP=8 across two nodes. There is also a minor checkpoint-resume issue with Adam: the optimizer state for layer 0’s residual is missing.

Please see the branch here for my local fix: https://github.com/pytorch/torchtitan/compare/main...elfiegg:torchtitan:fix/pp8-neighbor-p2p-metadata

PP workaround changed from 
```
metadata objects ─┐
mode-vote chain ──┼── full PP NCCL/order domain
some tensor P2P ──┘
```
to 
```
metadata objects ───── PP-replica Gloo group, CPU

edge 0↔1 tensors ───── dedicated two-rank NCCL group
edge 1↔2 tensors ───── dedicated two-rank NCCL group
...
edge 6↔7 tensors ───── dedicated two-rank NCCL group

static/dynamic decision ─ one aligned PP-wide all-reduce
```

* Metadata no longer touches GPU NCCL.
* Traffic on edge 6↔7 cannot be blocked behind unrelated edge 3↔4 traffic.
* Every tensor communicator has exactly the two ranks that must agree about its ordering.
* The vote is one operation all PP ranks enter together, rather than a serial pairwise handshake.
* Groups are eagerly created in deterministic order, satisfying PyTorch’s process-group creation requirements.

AttnRes makes dynamic metadata P2P becomes necessary → metadata uses full-PP NCCL communicator
  → PP8/1F1B exposes communicator initialization/ordering hazard, which we probably never stumbled upon before. 

QAT training already appears to function with the PR, but what's needed is a more thorough accuracy study to evaluate model quality, rather than functional correctness.
